resilience4j-circuitbreaker/src/test/java/io/github/resilience4j/circuitbreaker/CircuitBreakerFutureTest.java

```diff
@@ -0,0 +1,204 @@
+        assertThat(circuitBreaker.getMetrics().getNumberOfSuccessfulCalls()).isEqualTo(1);
+        assertThat(circuitBreaker.getMetrics().getNumberOfNotPermittedCalls()).isEqualTo(0);
+        assertThat(circuitBreaker.getState()).isEqualTo(CircuitBreaker.State.CLOSED);
+    }
+
+    @Test
+    public void shouldDecorateFutureAndThrowExecutionException() throws Exception {
+        CircuitBreaker circuitBreaker = CircuitBreaker.ofDefaults("testName");
+
+        final Future<String> future = executor.submit(() -> {
+            throw new RuntimeException("BAM!");
+        });
+
+        CircuitBreakerFuture<String> decoratedFuture = new CircuitBreakerFuture<>(circuitBreaker, future);
+
+        Throwable thrown = catchThrowable(() -> decoratedFuture.get());
+
+        assertThat(thrown).isInstanceOf(ExecutionException.class)
+                .hasCauseInstanceOf(RuntimeException.class);
+
+        assertThat(circuitBreaker.getMetrics().getNumberOfBufferedCalls()).isEqualTo(1);
+        assertThat(circuitBreaker.getMetrics().getNumberOfFailedCalls()).isEqualTo(1);
+        assertThat(circuitBreaker.getMetrics().getNumberOfSuccessfulCalls()).isEqualTo(0);
+        assertThat(circuitBreaker.getMetrics().getNumberOfNotPermittedCalls()).isEqualTo(0);
+        assertThat(circuitBreaker.getState()).isEqualTo(CircuitBreaker.State.CLOSED);
+    }
+
+    @Test
+    public void shouldDecorateFutureAndThrowTimeoutException() throws Exception {
+        CircuitBreaker circuitBreaker = CircuitBreaker.ofDefaults("testName");
+
+        final Future<String> future = executor.submit(() -> {
+            Thread.currentThread().sleep(10000);
+            return null;
+        });
+
+        CircuitBreakerFuture<String> decoratedFuture = new CircuitBreakerFuture<>(circuitBreaker, future);
+
+        Throwable thrown = catchThrowable(() -> decoratedFuture.get(5, TimeUnit.SECONDS));
+
+        assertThat(thrown).isInstanceOf(TimeoutException.class);
+
+        assertThat(circuitBreaker.getMetrics().getNumberOfBufferedCalls()).isEqualTo(1);
+        assertThat(circuitBreaker.getMetrics().getNumberOfFailedCalls()).isEqualTo(1);
+        assertThat(circuitBreaker.getMetrics().getNumberOfSuccessfulCalls()).isEqualTo(0);
+        assertThat(circuitBreaker.getMetrics().getNumberOfNotPermittedCalls()).isEqualTo(0);
+        assertThat(circuitBreaker.getState()).isEqualTo(CircuitBreaker.State.CLOSED);
+    }
+
+    @Test
+    public void shouldDecorateFutureAndCallerRequestCancelled() {
+        CircuitBreaker circuitBreaker = CircuitBreaker.ofDefaults("testName");
+
+        final Future<String> future = executor.submit(() -> "Hello World");
+        future.cancel(true);
+
+        CircuitBreakerFuture<String> decoratedFuture = new CircuitBreakerFuture<>(circuitBreaker, future);
+        Throwable thrown = catchThrowable(() -> decoratedFuture.get());
+
+        assertThat(thrown).isInstanceOf(CancellationException.class);
```

Comment:
This test seems to fail sometimes. Maybe you should use a Future mock instead.

Reply:
Ooh, yeah, if the operation completes before `cancel(true)` is called, it won't get cancelled.

Reply:
Thanks @clgroft , thats why it was failing intermittently, I will add `Thread.sleep()`